paperqa/llms.py

```diff
@@ -875,12 +907,56 @@ def clear(self) -> None:
         super().clear()
         self.texts = []
         self._embeddings_matrix = None
+        self._texts_filter = None
 
     def add_texts_and_embeddings(self, texts: Iterable[Embeddable]) -> None:
         super().add_texts_and_embeddings(texts)
         self.texts.extend(texts)
         self._embeddings_matrix = np.array([t.embedding for t in self.texts])
 
+    async def partitioned_similarity_search(
+        self,
+        query: str,
+        k: int,
+        embedding_model: EmbeddingModel,
+        partitioning_fn: Callable[[Doc], int],
+    ) -> tuple[Sequence[Embeddable], list[float]]:
+        scores: list[list[float]] = []
+        texts: list[Sequence[Embeddable]] = []
+
+        # duck typing for self.texts since they don't technically
+        # have to have .doc attributes
+        if self.texts and not hasattr(self.texts[0], "doc"):
+            logger.warning(
+                "Skipping partitioning: A partitioning_fn was provided "
+                "but the texts do not have a 'doc' attribute."
+            )
```

Comment:
Won't we hit this logged warning a ton?

To me this seems like a code smell for we should do something different

Reply:
I'm actually not sure how this would ever be hit - you'd need to build your own version of a `NumpyVectorStore`, and instantiate the objects in it outside of the PaperQA flow.

We could subclass the whole `NumpyVectorStore`, and change the texts type from `Embeddable` to `Texts`, but that would be the only change otherwise. So I relied on duck typing. 

To think about the scenario that would produce this warning -- you'd need to write a partitioning function which takes in `Doc` objects, and returns and `int`, then you build your own `NumpyVectorStore` and insert to it objects which don't have any `Doc` objects as attributes. That seems really fringe.

Reply:
Actually thought more about this — I think I’ll just add as an arg to all the requisite functions— then we skip this issue entirely. I’ll need to add the funcs to the tool initializations, which I’ll think through more.

Reply:
Re-structured this function, so we don't need this duck typing anymore!